pkg/tbtc/redemption.go

```diff
@@ -0,0 +1,201 @@
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"time"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+	"github.com/keep-network/keep-core/pkg/chain"
+)
+
+// RedemptionTransactionShape is an enum describing the shape of
+// a Bitcoin redemption transaction.
+type RedemptionTransactionShape uint8
+
+const (
+	// RedemptionChangeFirst is a shape where the change output is the first one
+	// in the transaction output vector. This shape makes the change's position
+	// fixed and leverages some SPV proof cost optimizations made in the Bridge
+	// implementation.
+	RedemptionChangeFirst RedemptionTransactionShape = iota
+	// RedemptionChangeLast is a shape where the change output is the last one
+	// in the transaction output vector.
+	RedemptionChangeLast
+)
+
+// RedemptionRequest represents a tBTC redemption request.
+type RedemptionRequest struct {
+	// Redeemer is the redeemer's address on the host chain.
+	Redeemer chain.Address
+	// RedeemerOutputScript is the output script the redeemed Bitcoin funds are
+	// locked to. This field is not prepended with the byte-length of the script.
+	RedeemerOutputScript []byte
+	// RequestedAmount is the TBTC amount (in satoshi) requested for redemption.
+	RequestedAmount uint64
+	// TreasuryFee is the treasury TBTC fee (in satoshi) at the moment of
+	// request creation.
+	TreasuryFee uint64
+	// TxMaxFee is the maximum value of the per-redemption BTC tx fee (in satoshi)
+	// that can be incurred by this request, determined at the moment of
+	// request creation.
+	TxMaxFee uint64
+	// RequestedAt is the time the request was created at.
+	RequestedAt time.Time
+}
+
+// redemptionFeeDistributionFn calculates the redemption transaction fee
+// distribution for the given redemption requests. The resulting list
+// contains the fee shares ordered in the same way as the input requests, i.e.
+// the first fee share corresponds to the first request and so on.
+type redemptionFeeDistributionFn func([]*RedemptionRequest) []int64
+
+// withRedemptionTotalFee is a fee distribution function that takes a
+// total transaction fee and distributes it evenly over all redemption requests.
+// If the fee cannot be divided evenly, the last request incurs the remainder.
+func withRedemptionTotalFee(totalFee int64) redemptionFeeDistributionFn {
+	return func(requests []*RedemptionRequest) []int64 {
+		requestsCount := int64(len(requests))
+		remainder := totalFee % requestsCount
+		feePerRequest := (totalFee - remainder) / requestsCount
+
+		feeShares := make([]int64, requestsCount)
+		for i := range requests {
+			feeShare := feePerRequest
+
+			if i == len(requests)-1 {
+				feeShare += remainder
+			}
+
+			feeShares[i] = feeShare
+		}
+
+		return feeShares
+	}
+}
+
+// assembleRedemptionTransaction constructs an unsigned redemption Bitcoin
+// transaction.
+//
+// Regarding input arguments, the walletMainUtxo parameter is mandatory and
+// must be set accordingly. The requests slice must contain at least one element.
```

Comment:
All parameters but `shape` are mandatory, right?

Reply:
Yeah. Added  that info to the fucntion's docstring: 55ffd641

Reply:
What do you think about removing this part:

> Regarding input arguments, the walletMainUtxo parameter is mandatory and must be set accordingly.

This is what caused by confusion. We say that `walletMainUtxo` is mandatory, then we describe requests slice, fee shares, and fee distribution without mentioning they are mandatory as well.

With the change from https://github.com/keep-network/keep-core/commit/55ffd641ba3aabe99818150640fea379f342041d, I think we can remove the mentioned sentence and this will be all less confusing.

Reply:
1bd586a7